Make saved tensors default hooks thread local

Varal7 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62909

This PR makes saved tensors default hooks thread local.
This allows using default hooks in a multithreaded context.

When hooks are never registered, the static bool `enabled` is left at false, so it should barely impact the performance of TLSGuard, see [benchmark](https://github.com/pytorch/pytorch/pull/62909#issuecomment-897673451).


Differential Revision: [D30165416](https://our.internmc.facebook.com/intern/diff/D30165416)

Varal7 committed:
Make saved tensors default hooks thread local

This PR makes saved tensors default hooks thread local.
This allows using default hooks in a multithreaded context.

[ghstack-poisoned]

Varal7 committed:
Update on "[for signal] Make saved tensors default hooks thread local"


This PR makes saved tensors default hooks thread local.
This allows using default hooks in a multithreaded context.

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30165416).

Varal7 committed:
Update on "[for signal] Make saved tensors default hooks thread local"


This PR makes saved tensors default hooks thread local.
This allows using default hooks in a multithreaded context.

Differential Revision: [D30165416](https://our.internmc.facebook.com/intern/diff/D30165416)

[ghstack-poisoned]

Varal7 committed:
Update on "Make saved tensors default hooks thread local"


This PR makes saved tensors default hooks thread local.
This allows using default hooks in a multithreaded context.

Differential Revision: [D30165416](https://our.internmc.facebook.com/intern/diff/D30165416)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30165416).

Varal7 commented on aten/src/ATen/record_function.cpp:
```diff
@@ -630,5 +630,4 @@ bool shouldRunRecordFunction(bool* pre_sampled) {
     return false;
   }
 }
-
```
woops

Varal7 commented on test/test_autograd.py:
```diff
@@ -9329,6 +9329,47 @@ def train_fn_grad(x):
+                warnings.simplefilter('always')
+                with warnings.catch_warnings(record=True) as w:
+                    y = x * x
+                    _self.assertEqual(len(w), 2)
```
This test is currently failing. 
It seems like the default hooks are not propagated through `DataParallel`.

albanD replied:
TLS is not propagate through DataParallel :(

albanD replied:
That's actually a good test to make sure that even though you set the hooks in the calling thread, they are not set in the DP worker threads :D

Varal7 replied:
Oh I see, I got confused because of the snippet below.
But now I'm realizing that GradMode is handled at the Python level in `torch/nn/parallel/parallel_apply.py` !

```python
import torch

class Model(nn.Module):
    def forward(self, x):
        y = x * x
        print(y.requires_grad)

x = torch.ones(5, 5, requires_grad=True)
model = nn.DataParallel(Model())

model(x)
with torch.no_grad():
    model(x)
```

Output:
```
True
True
False
False
```

albanD commented on torch/csrc/autograd/python_saved_variable_hooks.cpp:
```diff
@@ -45,39 +46,37 @@ namespace torch { namespace autograd {
-
   void PyDefaultSavedVariableHooks::set_hooks(py::function &pack_hook, py::function &unpack_hook) {
-    std::lock_guard<std::mutex> lock(mutex_);
+    PyObject *pack_hook_(nullptr), *unpack_hook_(nullptr);
```
Do you need to initialize these to nullptr? You actually set them below.

Varal7 replied:
If I don't do it here as well, I get a clang-tidy error. So another option would be to add a no-lint comment

Varal7 committed:
Update on "Make saved tensors default hooks thread local"


This PR makes saved tensors default hooks thread local.
This allows using default hooks in a multithreaded context.

Differential Revision: [D30165416](https://our.internmc.facebook.com/intern/diff/D30165416)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30165416).

Varal7 committed:
Update on "Make saved tensors default hooks thread local"


This PR makes saved tensors default hooks thread local.
This allows using default hooks in a multithreaded context.

Differential Revision: [D30165416](https://our.internmc.facebook.com/intern/diff/D30165416)

[ghstack-poisoned]

Varal7 commented on aten/src/ATen/SavedTensorHooks.cpp:
```diff
@@ -0,0 +1,34 @@
+}
+
+void SavedTensorDefaultHooks::set_hooks(PyObject* pack_hook, PyObject* unpack_hook) {
+  pack_hook_ = pack_hook;
```
Oh that's a good point, thanks!

Varal7 committed:
Update on "Make saved tensors default hooks thread local"


This PR makes saved tensors default hooks thread local.
This allows using default hooks in a multithreaded context.

Differential Revision: [D30165416](https://our.internmc.facebook.com/intern/diff/D30165416)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30165416).

Varal7 committed:
Update on "Make saved tensors default hooks thread local"


This PR makes saved tensors default hooks thread local.
This allows using default hooks in a multithreaded context.

Differential Revision: [D30165416](https://our.internmc.facebook.com/intern/diff/D30165416)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30165416).

Varal7 commented:
Benchmark using https://github.com/pytorch/pytorch/tree/master/benchmarks/instruction_counts:


![image](https://user-images.githubusercontent.com/8019486/129211751-7fde660a-18c0-48e5-a254-c9b8bf690946.png)
![image](https://user-images.githubusercontent.com/8019486/129211846-e929fad1-6dab-43bc-8bcc-8c61f19ff4c0.png)
![image](https://user-images.githubusercontent.com/8019486/129211860-8503144b-107f-4ddc-82c8-023da356da34.png)
![image](https://user-images.githubusercontent.com/8019486/129211874-4d597730-81de-4733-88d2-ac828195c4d4.png)
![image](https://user-images.githubusercontent.com/8019486/129211877-3fea7a86-c7ea-485c-a169-d9ee0305c6cc.png)
![image](https://user-images.githubusercontent.com/8019486/129211930-99b9b3d8-6a2d-4d86-8a4f-255744376dd5.png)
![image](https://user-images.githubusercontent.com/8019486/129211942-350b0cc1-9558-4edd-98c9-97696165acd4.png)
![image](https://user-images.githubusercontent.com/8019486/129211966-572d40fb-c695-487a-8d91-8d675e213874.png)

Full table: https://docs.google.com/spreadsheets/d/1XyiN2E8JR2Fc0G0V2rrJ3z7i8p-FDLRuC0mnwqAnbGY/edit#gid=321297804

albanD reviewed (approved):
LGTM

Varal7 committed:
Update on "Make saved tensors default hooks thread local"


This PR makes saved tensors default hooks thread local.
This allows using default hooks in a multithreaded context.

When hooks are never registered, the static bool `enabled` is left at false, so it should barely impact the performance of TLSGuard, see [benchmark](https://github.com/pytorch/pytorch/pull/62909#issuecomment-897673451).


Differential Revision: [D30165416](https://our.internmc.facebook.com/intern/diff/D30165416)

[ghstack-poisoned]

Varal7 commented on test/test_autograd.py:
```diff
@@ -9331,6 +9331,49 @@ def train_fn_grad(x):
+        def registers_hooks_for_each_thread():
+            with torch.autograd.graph.saved_tensors_hooks(pack, lambda x: x):
+                x = torch.ones(5, 5, requires_grad=True)
+                warnings.simplefilter('always')
```
Sorry, indeed it's not

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30165416).

Varal7 commented:
@pytorchbot retest this please
